src/peer_connection.cpp

```diff
@@ -5460,23 +5460,22 @@ namespace libtorrent
 	int peer_connection::wanted_transfer(int channel)
 	{
 		TORRENT_ASSERT(is_single_thread());
-		shared_ptr<torrent> t = m_torrent.lock();
 
-		const int tick_interval = (std::max)(1, m_settings.get_int(settings_pack::tick_interval));
+#if TORRENT_NO_FPU
+		int const tick_interval = (std::max)(1, m_settings.get_int(settings_pack::tick_interval) / 1000);
+#else
+		float const tick_interval = m_settings.get_int(settings_pack::tick_interval) / 1000.f;
+#endif
 
 		if (channel == download_channel)
 		{
-			return std::max((std::max)(m_outstanding_bytes
-				, m_recv_buffer.packet_bytes_remaining()) + 30
-				, int(boost::int64_t(m_statistics.download_rate()) * 2
-					* tick_interval / 1000));
+			return std::max((std::max)(m_outstanding_bytes, m_recv_buffer.packet_bytes_remaining())
+				+ 30, static_cast<int>(m_statistics.download_rate() * tick_interval));
```

Comment:
I think the idea behind the ``* 2`` is to have some slack, to not have run out entirely of quota by the time the next bandwidth request is made (i.e. stall) and also to allow connections that have just started up to increase their download rate, and not be stuck to the download rate they had at startup just because they don't get any more bandwidth assigned to them.

Do you see some improved results in practice removing the ``* 2``?

Reply:
Well, definitely web seed connections will benefit from this change as they will be sharing available quota better then before. For regular bt connections it is not that obvious.

Reply:
Also I was thinking of web connections to be restricted to send complete HTTP requests only. Now it is possible that, because of quota limitations, web connection will send partial HTTP request. I have been watching situation when web connection was sending 1 byte requests. This definitely must be avoided, especially when communicating to NGINX. The last one will append access.log with bad request records when it receives 1 byte.

Reply:
I suppose quota could be accrued in a counter here, or the bandwidth manager could get another parameter of the size of a "quanta". That way a bittorrent connection could ask for enough space to send whole messages at a time.

Reply:
the problem I fear of not using any multiplier here is that a peer with a low upload rate, will only get enough quota for a slow download rate, which means it will request quota for a low download rate. There's a feed-back loop that a peer could get stuck in.

I think there should be some multiplier, if not `2`, perhaps `* 3 / 2` or `* 4 / 3`.

Does that make sense?